### PR TITLE
[BEAM-2095] Made SourceRDD.Bounded iterator's hasNext idempotent

### PR DESCRIPTION
**Pull request by staslev**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`.
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by staslev** [comment]
Run Spark ValidatesRunner

**Comment by staslev** [comment]
Run Spark ValidatesRunner

**Comment by staslev** [comment]
retest this please

**Comment by staslev** [comment]
Run Spark ValidatesRunner

**Comment by staslev** [comment]
R: @aviemzur

File: runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java [hunk 1]
```diff
@@ -118,80 +121,132 @@ public Bounded(
...
+     *   Consequently, when the reader is closed, or when the reader has no further elements
+     *   available (i.e, {@link Source.Reader#advance()} returned <code>false</code>),
+     *   <code>hasNext</code> returns <code>false</code>.
+     * </p>
+     * <p>
+     *   Since this is a read-only iterator, an attempt to call <code>remove</code> will throw an
+     *   <code>UnsupportedOperationException</code>.
+     * </p>
+     */
+    static class ReaderToIteratorAdapter<T> implements Iterator<WindowedValue<T>> {
```
Comment (aviemzur) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java]: `@VisibleForTesting`?

**Comment by dhalperi** [comment]
Hey @staslev do we want this fix for the first stable release?

**Comment by staslev** [comment]
Hey @dhalperi , thanks for pinging.
I was contemplating it, but thought it would be nice to get a confirmation that this has indeed solved the reported issue. Dropped a line at https://issues.apache.org/jira/browse/BEAM-2095 the other day.

**Pull request closed without merging**